enterprise/trackedusers/data_collector.go

```diff
@@ -0,0 +1,192 @@
+package trackedusers
+
+//go:generate mockgen -destination=./mocks/mock_data_collector.go -package=mockdatacollector github.com/rudderlabs/rudder-server/enterprise/trackedusers DataCollector
+
+import (
+	"context"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"time"
+
+	migrator "github.com/rudderlabs/rudder-server/services/sql-migrator"
+
+	obskit "github.com/rudderlabs/rudder-observability-kit/go/labels"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+
+	"github.com/rudderlabs/rudder-go-kit/logger"
+
+	"github.com/lib/pq"
+
+	"github.com/spaolacci/murmur3"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/segmentio/go-hll"
+
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	txn "github.com/rudderlabs/rudder-server/utils/tx"
+)
+
+const (
+	idTypeUserID                       = "userID"
+	idTypeAnonymousID                  = "anonymousID"
+	idTypeUserIDAnonymousIDCombination = "userIDAnonymousIDCombination"
+
+	// changing this will be non backwards compatible
+	murmurSeed = 123
+)
+
+// DataCollector is interface to collect data from jobs
+type DataCollector interface {
+	CollectData(ctx context.Context, jobs []*jobsdb.JobT, tx *txn.Tx) error
+}
+
+type UniqueUsersCollector struct {
+	log         logger.Logger
+	hllSettings *hll.Settings
+	instanceID  string
+}
+
+func NewUniqueUsersCollector(log logger.Logger, dbConn string) (*UniqueUsersCollector, error) {
+	dbHandle, err := sql.Open("postgres", dbConn)
+	if err != nil {
+		panic(err)
+	}
+	dbHandle.SetMaxOpenConns(1)
+
+	m := &migrator.Migrator{
+		Handle:                     dbHandle,
+		MigrationsTable:            "tracked_users_reports_migrations",
+		ShouldForceSetLowerVersion: config.GetBool("SQLMigrator.forceSetLowerVersion", true),
+	}
+	err = m.Migrate("tracked_users")
+	if err != nil {
+		return nil, fmt.Errorf("could not run tracked_users_reports migrations: %w", err)
+	}
```

Comment:
I would avoid running a migration as part of new.

We could instead:
1. Expose a separate method for running migration. Expecting users of the method to call it. 
2. or use sync.once on `CollectData` to ensure it run the first time CollectData call was made, but makes error handling a bit more complicated.

We should pass a context to migration as well.

Reply:
migrator currently do not takes context as a param, I can create a new task for adding support for context while running migrations.